### ISSUE TITLE
CAM features

### ISSUE THREAD

**Issue by Harvie** [codelv/declaracad#9]
Hello, have you been thinking about CAM features for CNC milling? Currently there's no declarative CAM and i think it would be cool to have that. CAM works best when it's integrated directly in CAD. So i think it would make sense to have declarative CAM in DeclaraCAD.

It would basicaly mean that you would declare shape of stock material to begin with, types of milling operation for different parts of DeclaraCAD model and it will then generate toolpaths for milling machine that would result into making that part. Like this:

![image](https://user-images.githubusercontent.com/276504/42877132-f631d5b4-8a88-11e8-89ff-7862db0f44ba.png)

**Comment by frmdstryr** [comment]
Haven't thought about it as I was mainly using this for modeling, but I think you have a very good idea.  

I don't believe OpenCascade's [CAM library](https://www.opencascade.com/content/cam-library) is open source.  Do you know of any open source libraries that provide similar functionality? 

Could you go into more detail on how you envision this being implemented? Perhaps some pseudocode?

**Comment by Harvie** [comment]
One year ago i've already started discussion on this with OpenSCAD people. https://github.com/openscad/openscad/issues/2017

They are not very enthusianistic about it. Basicaly because they have not enogh capacity to implement this right now.

What i figured out during this discussion is that it might be quite easy to start with following:

1.) implement structure to represent and visualize 3D paths in DeclaraCAD
2.) add way to import g-code to the internal representation of 3D toolpaths mentioned in 1.)
3.) add possibility to combine this 3D toolpath with any shape, so the shape follows the toolpath. That way you would get solid object which represents space which will be milled by the mill of any shape. You can then compute difference of stock material and this shape. BAM: you have g-code simulator.

After this the DeclaraCAD would understand the g-code and it would be easy to implement real CAM functionality on top of that. After all you can do basic 3D CAM by slicing the mesh and running 2D operations on the slices. You can even have external script for this, which will take model from DeclaraCAD and feed the g-code back to it using g-code import function.

**Comment by Harvie** [comment]
These are so cool news!!! Really glad to hear that.

Parsing g-code in python is fairly easy to do.
Is the "Wire" really 3D? Because in g-code you can have arc in XY plane, which have different Z height at start and end. In other words a 3D helix.

**Comment by Harvie** [comment]
BTW you can simulate milling by making difference of the "pipe" from the stock material.
But if you use small sphere rather than model of milling bit you can simulate 3D printing :-) Of course when you simulate additive manufacturing you don't need to subtract anything after making the "Pipe"...

**Comment by frmdstryr** [comment]
> Is the "Wire" really 3D? 

Yes it maps to https://www.opencascade.com/doc/occt-6.9.0/refman/html/class_b_rep_builder_a_p_i___make_wire.html . All the points are 3d.

I think BSplines and Beziers need added yet.

**Comment by Harvie** [comment]
I think most common opensource 3 axis CNC mill controller nowadays is GRBL, which only understands g-code with LINES (G0,G1) and ARCS (G2,G3). Also bCNC 2D CAM which i use handles everything using lines and arcs and completely ignores beziers and splines. They just approximate compound shapes using lots of short G1 lines. So i can say that you don't really need to support them.

I am not sure about 3D printing world (Marlin), i think they might use splines, but it does not really make sense, because they slice STLs anyway. And STLs are triangulated, so you always get just lines. Unless you artificialy replace resulting series of lines with splines.

So don't worry. All we need to begin are lines and (helical) arcs.

**Comment by frmdstryr** [comment]
I think some optimizations may need done for this. I was able to load a few sample gcode files from https://www.cnccookbook.com/g-code-examples/. I don't think the performance is not good enough, it takes a second or two to load on my laptop, and adding a Pipe and Cut will be worse. For things like this I think more direct apis should be added to load GCode to using OpenCascade primitives directly without python overhead.

![declaracad-gcode-test](https://user-images.githubusercontent.com/380158/42943695-7c57f846-8b31-11e8-80e3-dc4644c5fbe9.gif)


Here's the code:

```python

import re
from declaracad.occ.api import (
    Part, Wire, Circle, Ellipse, Segment, Point, Polygon
)
from enaml.core.api import Looper, Include

def split_gcode_commands(path):
    # Very simple parser
    cmds = []
    x, y, z = 0, 0, 0
    with open(path) as f:
        for line in f:
            cmd = re.split(r';|\(', line)[0]
            if cmd:
                cmd = {c[0]:float(c[1:]) for c in cmd.split()}
                x = cmd.get('X', x)
                y = cmd.get('Y', y)
                z = cmd.get('Z', z)
                cmd['P'] = (x, y, z)
                cmds.append(cmd)
    # Split into separate wires
    wires = []
    for cmd in cmds:
        G = cmd.get('G')
        if G == 0:
            wires.append([])
        elif G == 1:
            wires[-1].append(cmd)
        # TODO: Arcs
    wires = [w for w in wires if len(w)>1]
    return wires
            
enamldef Assembly(Part):
    name = "Draw Example"
    Looper:
        iterable << split_gcode_commands('boomerangv4.ncc')
        Wire:
            Segment:
                Looper:
                    iterable << loop_item
                    Point:
                        attr cmd = loop_item
                        position = cmd.get('P')
```

**Comment by Harvie** [comment]
This is free library usefull for CAM operations: https://github.com/aewallin/opencamlib

**Comment by frmdstryr** [comment]
How large is a typical "real world" gcode file?

**Comment by Harvie** [comment]
It really depends on what are you doing, how precise surface you need and how efficient is the g-code generator (CAM)... The boomerang you've used is about 8000 lines.  I think you can do a lot with 8000 to 16000 lines. I have some g-code files around 100 lines of code, also i have some files of 300k lines. It really depends. I can say that lots of the programs already have troubles with such big files. Nobody expects this to be processed immediately. So i guess if you get 10000 lines processed in few seconds, it will be quite usable and some optimalizations can be done if this proves usefull. Especialy the pipe does not have to be realtime. "CAMotics" g-code simulator takes some time to process g-code back to the mesh even with poor accuracy. The pipe algorithm would produce much more accurate results, so it's perfectly ok if it will take few minutes for big files.

**Comment by frmdstryr** [comment]
It seems the pipe algorithm isn't working exactly how I had intended for this. See this https://stackoverflow.com/questions/47163841/pythonocc-opencascade-create-pipe-along-straight-lines-through-points-profile. 

![declaracad-gcode-pipe-test](https://user-images.githubusercontent.com/380158/43047304-5e936fec-8da3-11e8-9108-87dfa15d462d.gif)

I think the profile will need extruded for each section which will make arcs difficult, unless opencascade has an api to "extrude" a shape along a path.

**Comment by frmdstryr** [comment]
Well there's still a lot to learn and improve on... but I've managed to get it to properly show the output a cut path: https://declaracad.com/docs/cam/simulation/

**Comment by Harvie** [comment]
@frmdstryr impressive!

**Comment by frmdstryr** [comment]
It will now attempt to load and render the gcode paths if a file is opened.

![declaracad-gcode-view](https://user-images.githubusercontent.com/380158/92113073-c8663f00-edbc-11ea-8d98-6499c8555d7c.png)
